components/calendar/src/types.rs

```diff
@@ -41,37 +51,99 @@ pub struct FormattableYear {
 
     /// The year number in the current era (usually 1-based).
     pub number: i32,
+}
 
-    /// The year in the current cycle for cyclic calendars (1-indexed)
-    /// can be set to `None` for non-cyclic calendars
-    ///
-    /// For chinese and dangi it will be
-    /// a number between 1 and 60, for hypothetical other calendars it may be something else.
-    pub cyclic: Option<NonZeroU8>,
+/// Representation of a year as needed for formatting
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[non_exhaustive]
+pub struct FormattableYear {
+    /// The "extended year", typically anchored with year 1 as the year 1 of either the most modern or
+    /// otherwise some "major" era for the calendar
+    pub extended_year: i32,
+    /// The rest of the details about the year.
+    pub kind: FormattableYearKind,
+}
 
-    /// The related ISO year. This is normally the ISO (proleptic Gregorian) year having the greatest
-    /// overlap with the calendar year. It is used in certain date formatting patterns.
+/// The type of year: Calendars like Chinese don't have an era and instead format with cyclic years.
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[allow(clippy::exhaustive_enums)] // Only two kinds of years
+pub enum FormattableYearKind {
+    /// An Era and a numeric year in that era.
+    EraYear(Era, i32),
+    /// A cyclic year, and the related ISO year
     ///
-    /// Can be `None` if the calendar does not typically use `related_iso` (and CLDR does not contain patterns
-    /// using it)
-    pub related_iso: Option<i32>,
+    /// Knowing the cyclic year is typically not enough to pinpoint a date, however cyclic calendars
+    /// don't typically use eras, so disambiguation can be done by saying things like "Year 甲辰 (2024)"
+    Cyclic(NonZeroU8, i32),
 }
 
-impl FormattableYear {
+impl YearInfo {
     /// Construct a new Year given an era and number
-    ///
-    /// Other fields can be set mutably after construction
-    /// as needed
-    pub fn new(era: Era, number: i32, cyclic: Option<NonZeroU8>) -> Self {
+    pub fn new(extended_year: i32, era: TinyStr16, number: i32) -> Self {
         Self {
-            era,
+            extended_year,
+            era: Era(era),
             number,
-            cyclic,
-            related_iso: None,
         }
     }
 }
 
+impl FormattableYear {
+    /// Construct a new Year given an era and number
+    pub fn new_era(extended_year: i32, era: TinyStr16, number: i32) -> Self {
```

Comment:
Nit: I'm not generally a fan of functions with positional arguments, especially when the positional arguments can be confused. For example, it is easy to mix up `extended_year` and `number` here.

Is constructing these from enums/structs too awkward?

Reply:
Somewhat, but also part of the goal was to be able to keep tweaking these without needing to change callsites each time